sys/log: Add optional support for sector bookmarks to optimize reading logs

vrahane opened the pull request:
- This adds sector bookmarks for reading optimization, earlier only absolute bookmarks were supported
- Retain older absolute bookmarks behavior. Added a syscfg for specifying number of absolute bookmarks.
- Added a syscfg to optionally enable sector bookmarks.
- CLI: Add '-i' option in log_shell for specifying index where log read should start
- CLI: Add '-b' option in log_shell for reading bookmarks

For mixed bookmarks:
- Sector bookmarks are from `0` -> `sector_cnt` of the flash area in the array.
- Non-sector/Absolute bookmarks are from `sector_cnt `-> `LOG_FCB_NUM_ABS_BMARKS` (2 by default)

Several tests were performed:
- [x] Reboot test, bookmarks should get reset
- [x] New log entries populate the bookmarks, specially for sector boundaries
- [x] Test absolute bookmarks behavior, both with mixed bookmarks (sector + absolute) as well as just absolute bookmarks with sector bookmarks disabled

kasjer reviewed (commented):
- Code will not build for FCB2
- It's not clear how to choose size for bookmarks array
- If suggested size is around number of sector of the FCB flash area, RAM usage can be an issue. For 384 sectors' FCB RAM usage is over 12KB
- There are places that use code and RAM space even if no bookmarks/sector-bookmarks are enabled

vrahane commented on sys/log/full/syscfg.yml:
```diff
@@ -179,6 +179,18 @@ syscfg.defs:
             the global index to be sequentially increasing for persisted logs.
         value: 0
 
+    LOG_FCB_NUM_ABS_BMARKS:
```
Just trying to reduce the length of the name since it was getting too big but I guess I could make all of them `_BMARKS`.

vrahane replied:
What about `LOG_FCB_BOOKMARKS` they cannot be changed ? If thats the case I think its best to keep this one syscfg as an exception. Do you agree ?

vrahane replied:
I guess you are right, 3 characters more isn’t that bad. Will change everything to be BOOKMARKS.

vrahane replied:
Changed.

vrahane commented on sys/log/full/include/log/log_fcb.h:
```diff
@@ -39,6 +39,8 @@ struct log_fcb_bmark {
 #endif
     /* The index of the log entry that the FCB entry contains. */
     uint32_t lfb_index;
+    /* Is this a sector boundary bookmark */
+    bool lfb_sect_bmark;
```
Left over from a previous version of the implementation. will remove. Thanks.

vrahane commented on sys/log/full/src/log_fcb_bmark.c:
```diff
@@ -70,65 +118,246 @@ log_fcb_clear_bmarks(struct fcb_log *fcb_log)
     }
 
-    bset->lfs_bmarks[bset->lfs_next] = (struct log_fcb_bmark) {
+    memcpy(&bset->lfs_bmarks[pos + 1], &bset->lfs_bmarks[pos],
```
I guess for x86_64 and i386 builds it might be an issue. For ARM, it seems fine though. Either ways will make the change.

vrahane commented on sys/log/full/include/log/log_fcb.h:
```diff
@@ -49,11 +51,17 @@ struct log_fcb_bset {
 
     /** The index where the next bookmark will get written. */
-    int lfs_next;
+    uint32_t lfs_next;
```
Will remove it.

vrahane commented on sys/log/full/src/log_shell.c:
```diff
@@ -39,7 +39,7 @@
 #include "log_cbor_reader/log_cbor_reader.h"
 
-void log_console_print_hdr(const struct log_entry_hdr *hdr);
+void log_console_print_hdr(struct log_entry_hdr *hdr);
```
That was an oversight probably but the const's were removed mainly because on our MCU we were seeing lower performance because the flash is external, not particularly with this code but in general. So, trying to reduce const usage wherever possible. Will change this back if you think its an issue.

vrahane replied:
Got it, the walk body function’s function pointer has a const pointer as an argument. So, changing this has no effect unless everything is changed. Well, that’s too big of a change for this PR. Will change this back.

vrahane commented on sys/log/full/src/log_fcb_bmark.c:
```diff
@@ -70,65 +121,247 @@ log_fcb_clear_bmarks(struct fcb_log *fcb_log)
-    if (bset->lfs_cap == 0) {
-        return;
+    if (pos == bset->lfs_cap - 1) {
+        goto add;
```
Never used it but heard stories, but as for this code will move up the statement.

vrahane commented on sys/log/full/src/log_fcb_bmark.c:
```diff
@@ -70,65 +121,247 @@ log_fcb_clear_bmarks(struct fcb_log *fcb_log)
+        }
+    } else
+#endif
+    if (!en_sect_bmark) {
+        for (i = 0; i < bset->lfs_non_s_size; i++) {
+            if (index == bset->lfs_bmarks[i].lfb_index) {
+                /* If index matches, no need to replace */
+                return;
+            }
+        }
     }
```
It was correct before, in an effort to make the code better I messed it up. Will fix it. Thanks, good catch.

vrahane replied:
This is actually correct because this API only replaces non-sector bookmarks. I will rename it so that there is no confusion going forward.

vrahane commented on sys/log/full/src/log_fcb_bmark.c:
```diff
@@ -70,65 +121,247 @@ log_fcb_clear_bmarks(struct fcb_log *fcb_log)
-log_fcb_add_bmark(struct fcb_log *fcb_log, const struct fcb_entry *entry,
-                  uint32_t index)
+static void
+log_fcb_insert_bmark(struct fcb_log *fcb_log, struct fcb_entry *entry,
```
The const's were removed mainly because on our MCU we were seeing lower performance because the flash is external, not particularly with this code but in general. So, trying to reduce const usage wherever possible. Will change this back if you think its an issue.

vrahane commented on sys/log/full/src/log_fcb2.c:
```diff
@@ -476,12 +485,13 @@ log_fcb2_walk(struct log *log, log_walk_func_t walk_func,
     struct fcb_log *fcb_log;
     struct fcb2_entry loc;
     int rc;
+    int min_diff = -1;
```
`min_diff` seemed fine because it is the minimum difference in distance between the entry and the bookmark when a match is found, otherwise it stays -1. If an exact match is found it's 0 which also fits its purpose and name. Do you have any other name in mind ?

vrahane replied:
I did not add min_diff it was already there. I just made it available. I changed it little bit so that `-1` indicates no match found, 0 means an exact match found and non-zero means match found at some distance.

vrahane replied:
Here if where it was there already: https://github.com/apache/mynewt-core/pull/3365/files#diff-0ee3178fe8ba7dd0ec1c36b8bf4c99c1a060d64c0f1056dfaecfcf3f36261b22L91

vrahane commented on sys/log/full/src/log_fcb.c:
```diff
@@ -108,6 +110,7 @@ fcb_walk_back_find_start(struct fcb *fcb, struct log *log, struct log_offset *lo
  * ignored; the "index" field is used instead.
  *
  * XXX: We should rename "timestamp" or make it an actual timestamp.
+ * If bmark is found, fill up min_diff.
```
Since it was a static function, I didn't explain it much but I explained it for a public api here: https://github.com/apache/mynewt-core/pull/3365/files#diff-a8268376783e79148c79b3f1117fc7f0bd35d55ff075dfc20779e8063496e561R165

vrahane commented on sys/log/full/src/log_fcb2.c:
```diff
@@ -179,6 +180,14 @@ log_fcb2_start_append(struct log *log, int len, struct fcb2_entry *loc)
+        /* The FCB needs to be rotated, reinit previously allocated
+         * bookmarks
+         */
+        log_fcb_init_bmarks(fcb_log, fcb_log->fl_bset.lfs_bmarks,
```
Wonder how the unit tests passed. I did fix it earlier but since I was debugging on a separate machine, I might have made a mistake while rebasing.

vrahane commented on sys/log/full/include/log/log_fcb.h:
```diff
@@ -39,21 +39,31 @@ struct log_fcb_bmark {
     /** Array of bookmarks. */
     struct log_fcb_bmark *lfs_bmarks;
 
+    bool lfs_en_sect_bmarks;
```
Yes, it is the same. Was trying to reduce the length of the variables where ever possible. non_sect would be too big so, made it non_s. What do you think ? Is _s_ on its own is too vague I feel.

vrahane replied:
Changing it to non_sect.

vrahane replied:
Will just change it to _sect_.

vrahane commented on sys/log/full/src/log_fcb_bmark.c:
```diff
@@ -22,17 +22,68 @@
+        return -1;
+    }
+
+    for (i = 0; i < fcb_log->fl_fcb.f_sector_cnt; i++) {
```
Yeah, I wanted to keep the API as much backwards compatible as possible without adding more complexity. But if needed I could add a separate `log_fcb_init_sect_bmarks()` API, then the user would have to utilize both APIs instead of having just one. Do you have any other suggestions how to make this better ? I could also have return codes indicating sector bookmarks did not fit. Asserting is also an option, better than an array overrun I guess.

vrahane replied:
Sector based bookmarks is a new functionality, old functionality does not get affected unless this is enabled specifically. I do not think this will break anything that is used currently by anyone. If you have a specific suggestion I can try to change it that way else I will just add a comment in the init code. does that sound fine ?

vrahane replied:
Added a comment. Please let me know if its good enough.

vrahane commented on sys/log/full/src/log_fcb_bmark.c:
```diff
@@ -70,65 +121,247 @@ log_fcb_clear_bmarks(struct fcb_log *fcb_log)
+     */
     for (i = 0; i < fcb_log->fl_bset.lfs_size; i++) {
         bmark = &fcb_log->fl_bset.lfs_bmarks[i];
+#if MYNEWT_VAL(LOG_FCB)
+        if (!fcb_log->fl_bset.lfs_bmarks[i].lfb_entry.fe_area) {
+            /* Previous closest bookmark is the closest one */
+            break;
+        }
+#elif MYNEWT_VAL(LOG_FCB2)
+        if (!fcb_log->fl_bset.lfs_bmarks[i].lfb_entry.fe_range) {
+            /* Previous closest bookmark is the closest one */
+            break;
+        }
+#endif
```
For sector bookmarks, the number of bookmarks are going to be equal to the sector_cnt and after that by default 2 bookmarks would be present for non-sector bookmarks. Now, if the sector bookmarks are not present, then only the non-sector bookmarks would be in the array. Also, sector bookmarks get initialized when bookmarks get initialized, so, there is no way the sector bookmarks are not going to be filled.

vrahane replied:
Fixed

vrahane commented on sys/log/full/src/log_fcb_bmark.c:
```diff
@@ -70,65 +121,247 @@ log_fcb_clear_bmarks(struct fcb_log *fcb_log)
+    (void)en_sect_bmark;
+#if MYNEWT_VAL(LOG_FCB_SECTOR_BOOKMARKS)
+    if (en_sect_bmark) {
+        for (i = 0; i < bset->lfs_size; i++) {
+            if (index > bset->lfs_bmarks[i].lfb_index) {
+                pos = i;
+                break;
+            } else {
+                pos = i + 1;
+            }
+        }
```
What do you mean ? pos tracks i. If the bookmark we are adding has a higher index than the index in the bookmark at i, pos will become that value. If index is <= index of bookmark at i, we keep incrementing pos and iterating forward until we find the right i in which case we break and come out of the loop. How is setting pos to `bset->lfs_size` even correct ? If you try it out, you will see it is correct the way it is. I have actually written the sorting code separately and then added here after testing it. If you feel doubtful can you please try it out ?

vrahane replied:
Go it, will change it.

vrahane commented on sys/log/full/src/log_fcb_bmark.c:
```diff
@@ -70,65 +121,247 @@ log_fcb_clear_bmarks(struct fcb_log *fcb_log)
+            bset->lfs_next_non_s = (bset->lfs_next_non_s + 1) %
+                                   MYNEWT_VAL(LOG_FCB_NUM_ABS_BMARKS);
+        } else {
+            if (!bset->lfs_non_s_size) {
+                /* First non-sector bmark position */
+                bset->lfs_next_non_s = pos;
+            }
```
What about:
```
323             if (!bset->lfs_non_s_size) {
324                 /* First non-sector bmark position */
325                 bset->lfs_next_non_s = pos;
326             }
```

vrahane replied:
I see, yup, I think while optimizin the code I messed it up, it was this before:
```
    } else {
        if (bset->lfs_non_s_size >= MYNEWT_VAL(LOG_FCB_NUM_ABS_BMARKS)) {
            /* Replace oldest non-sector bmark */
            log_fcb_replace_bmark(fcb_log, entry, index, sect_bmark,
                                  bset->lfs_next_non_s +
                                  fcb_log->fl_fcb.f_sector_cnt);
            MODLOG_DEBUG(LOG_MODULE_DEFAULT, "replace bmark index: %u, pos: %u, \n",
                         (unsigned int)index,
                         (unsigned int)bset->lfs_next_non_s + fcb_log->fl_fcb.f_sector_cnt);
            bset->lfs_next_non_s = (bset->lfs_next_non_s + 1) %
                                   MYNEWT_VAL(LOG_FCB_NUM_ABS_BMARKS);
        } else {
            log_fcb_replace_bmark(fcb_log, entry, index, sect_bmark,
                                  fcb_log->fl_fcb.f_sector_cnt +
                                  bset->lfs_non_s_size);
            MODLOG_DEBUG(LOG_MODULE_DEFAULT, "replace bmark index: %u, pos: %u, \n",
                         (unsigned int)index,
                         (unsigned int)fcb_log->fl_fcb.f_sector_cnt + bset->lfs_non_s_size);
            if (!bset->lfs_non_s_size) {
                /* First non-sector bmark position */
                bset->lfs_next_non_s = pos;
            }
            bset->lfs_non_s_size++;
            bset->lfs_size++;
        }
  ```
  Which was not optimized but worked. Will fix it. Thanks.

vrahane commented on sys/log/full/src/log_fcb_bmark.c:
```diff
@@ -61,74 +112,246 @@ log_fcb_rotate_bmarks(struct fcb_log *fcb_log)
 
-    if (bset->lfs_cap == 0) {
-        return;
+    if (pos != bset->lfs_cap - 1) {
+        memmove(&bset->lfs_bmarks[pos + 1], &bset->lfs_bmarks[pos],
+                sizeof(bset->lfs_bmarks[0]) * (bset->lfs_size - pos - 1));
     }
```
It wasn't crashing before, let me take a look. Is this because of memmove ? memcpy was working just fine.

kasjer replied:
It was working only because optimized implementation treated size as signed and did not try to do anything for negative value. Size is computed wrong when lfs_size is 0

vrahane replied:
fixed

vrahane commented on sys/log/full/src/log_shell.c:
```diff
@@ -172,11 +174,18 @@ shell_log_dump_cmd(int argc, char **argv)
     int rc;
 
     clear_log = false;
+    (void)dump_bmarks;
+    (void)bmarks;
+    (void)bmarks_size;
```
I still see build errors. If you dont have an issue, I am going to keep this.

vrahane commented on sys/log/full/src/log_fcb2.c:
```diff
@@ -47,14 +47,16 @@ static int log_fcb2_rtr_erase(struct log *log);
  * The "index" field corresponds to a log entry index.
  *
  * If bookmarks are enabled, this function uses them in the search.
+ * For min_diff, if bookmark was found, fill it up with the difference
+ * else it will return -1.
```
I will give it another shot, if you don't like that one, you can change it yourself or you can suggest something specific. It is a static function, so, no external users will use it anyway.

vrahane replied:
Done.

kasjer commented:
@vrahane stop marking comments as resolved, let me decide if I agree with resolution

vrahane commented:
@kasjer Once I have addressed them, I resolve them. You can just click on them one by one and check. They don't go anywhere. I dont mark them as unresolved unless I have addressed them. Also, please don't re-review yet. I am still seeing non-sector bookmarks having an issue after I addressed all the review comments. I need to debug that.

kasjer commented:
> @kasjer Once I have addressed them, I resolve them. You can just click on them one by one and check. They don't go anywhere. I dont mark them as unresolved unless I have addressed them. Also, please don't re-review yet. I am still seeing non-sector bookmarks having an issue after I addressed all the review comments. I need to debug that.

When you addressed them it clearly does not mean they are resolved them

vrahane commented:
Ok, whichever you feel is not done, please un-resolve it.

vrahane commented:
@kasjer I have tried to address all comments. Please let me know what you think ?

kasjer commented on sys/log/full/selftest/fcb_bookmarks/src/log_test_fcb_bookmarks_util.c:
```diff
@@ -231,7 +231,7 @@ ltfbu_init(const struct ltfbu_cfg *cfg)
 
     if (cfg->bmark_count > 0) {
-        log_fcb_init_bmarks(&ltfbu_fcb_log, ltfbu_bmarks, cfg->bmark_count);
+        log_fcb_init_bmarks(&ltfbu_fcb_log, ltfbu_bmarks, cfg->bmark_count, false);
```
This is the only place in mynewt repository where `log_fcb_init_bmarks()` is called. When I try to put similar code (except last argument is true) in actual code system crashes.
Is there a restriction on when this function should be called?

vrahane replied:
Well, there is no restrictions as such, I called it on the fcb from the slinky app and that seemed to work. Where did you call it from ?

kasjer replied:
This is test function called from main after
```c
int
test_log_init(void)
{
    const struct flash_area *fa;
    struct fcb *fcbp;
    int sector_count;
    int rc;

    if (flash_area_open(FLASH_AREA_TEST_LOG, &fa)) {
        return SYS_EUNKNOWN;
    }

    flash_area_to_sectors(FLASH_AREA_TEST_LOG, &sector_count, NULL);
    test_log_sectors = calloc(sector_count, sizeof(*test_log_sectors));
    flash_area_to_sectors(FLASH_AREA_TEST_LOG, &sector_count, test_log_sectors);
    test_log_fcb.fl_entries = 0;
    fcbp = &test_log_fcb.fl_fcb;

    fcbp->f_magic = 0x7EADBADF;
    fcbp->f_version = g_log_info.li_version;
    fcbp->f_sector_cnt = sector_count - 1;
    fcbp->f_scratch_cnt = 1;
    fcbp->f_sectors = test_log_sectors;

    rc = fcb_init(fcbp);
    if (rc) {
        flash_area_erase(fa, 0, fa->fa_size);
        rc = fcb_init(fcbp);
        if (rc) {
            return rc;
        }
    }

#if MYNEWT_VAL_LOG_FCB_SECTOR_BOOKMARKS
    int bookmark_count = fcbp->f_sector_cnt + MYNEWT_VAL_LOG_FCB_NUM_ABS_BOOKMARKS;
    test_log_bookmarks = malloc(bookmark_count * sizeof(*test_log_bookmarks));

    log_fcb_init_bmarks(&test_log_fcb, test_log_bookmarks, bookmark_count, true);
#endif

    rc = log_register("test_log", &test_log, &log_fcb_handler,
                      &test_log_fcb, LOG_SYSLEVEL);

    return rc;
}
```
Order of calls based on the only example found in unit tests except unit tests call to `log_fcb_init_bmarks()` has **false** as last argument:
- fcb_init()
- log_fcb_init_bmarks() <- crashes inside
- log_register()

kasjer commented on sys/log/full/src/log_fcb_bmark.c:
```diff
@@ -61,74 +132,258 @@ log_fcb_rotate_bmarks(struct fcb_log *fcb_log)
 
-    if (bset->lfs_cap == 0) {
-        return;
+    if (bset->lfs_size && bset->lfs_size > pos && pos != (bset->lfs_cap - 1)) {
+        memmove(&bset->lfs_bmarks[pos + 1], &bset->lfs_bmarks[pos],
+                sizeof(bset->lfs_bmarks[0]) * (bset->lfs_size - pos));
```
This code keeps sector bookmarks sorted by index.
It does not look like any other place in the code takes advantage of this order. When bookmarks are search for match this order is not utilized.
Maybe there is not need to move chunk of memory every time when sector bookmark is inserted which seems to be the case since sector are scanned from the oldest one. So every sector bookmark result in memmove with increasing size.

vrahane replied:
This was done because earlier absolute bookmarks could actually be in between the other bookmarks based on the index. Even now, insertion happens as pos 0 but, we could optimize it so that there is no memmove happening. Let me try it.

vrahane replied:
Ok, I did optimize the code quite a bit. Please take a look. As per the review comment I have removed memmove.

andrzej-kaczmarek commented on sys/log/full/include/log/log_fcb.h:
```diff
@@ -103,19 +111,34 @@ struct fcb_log {
+ *        If sector bookmarks are enabled, buffer should be big enough
+ *        to accomodate bookmarks for the entire flash area that is allocated
+ *        for the FCB log, i,e; sizeof(struct log_fcb_bmark) *
+ *        my_log.fl_fcb.f_sector_cnt + MYNEWT_VAL(LOG_FCB_NUM_ABS_BOOKMARKS)
```
It can't stay like this. The function should return an error if buffer is too small and not just crash later on. But actually it would be better if it can work with smaller buffers as well and e.g. place bookmark every n-th sector or just skip bookmarks that don't fit in the buffer.

vrahane replied:
That is possible but it complicates the code further. Just failing is a better approach. The user should know what they are doing. What if they have a bigger buffer ? Then we will have to handle that too. I would like to keep it ab it restricted in its usage.

vrahane replied:
Ok, let me give it a shot.

vrahane replied:
@andrzej-kaczmarek  Ok, I have added support for this. Please take a look. I am going to exercise it little bit more to make sure there are no edge cases I have forgotten.

vrahane commented:
@kasjer  and @andrzej-kaczmarek build failures are unrelated.

kasjer commented:
This time I did not look at the code just tested it on the entirely full fcb log on 1MB spi flash with 6 absolute bookmarks.
It seems that after 6 walks sector bookmark are no longer used
here are some output from log console
```
compat> log -b test_log
0: index:0 entry: 20007d28 fa: 0 fe_elem_off: 0
017947 1: index:68837 entry: 20007d44 fa: 2000713c fe_elem_off: 2c
...
018263 250: index:15108 entry: 20009880 fa: 20007184 fe_elem_off: a5
018265 251: index:13154 entry: 2000989c fa: 20007178 fe_elem_off: 2b
018266 252: index:11764 entry: 200098b8 fa: 2000716c fe_elem_off: 2b
018267 253: index:10262 entry: 200098d4 fa: 20007160 fe_elem_off: 2a
018269 254: index:8819 entry: 200098f0 fa: 20007154 fe_elem_off: 8
018270 255: index:0 entry: 2000990c fa: 0 fe_elem_off: 0
018271 256: index:0 entry: 20009928 fa: 0 fe_elem_off: 0
018272 257: index:0 entry: 20009944 fa: 0 fe_elem_off: 0
018273 258: index:0 entry: 20009960 fa: 0 fe_elem_off: 0
018275 259: index:0 entry: 2000997c fa: 0 fe_elem_off: 0
018275 260: index:0 entry: 20009998 fa: 0 fe_elem_off: 0
```
Last 6 elements are for absolute bookmarks I guess
Now command to measure time of walking 2 entries from index (visible in sector bookmarks)
```
compat> log -i 10262 -n 2 -t test_log
253519 Log test_log 2 entries walked in 83 ms
compat> log -b test_log
362773 0: index:0 entry: 20007d28 fa: 0 fe_elem_off: 0
362774 1: index:68837 entry: 20007d44 fa: 2000713c fe_elem_off: 2c
...
363093 252: index:11764 entry: 200098b8 fa: 2000716c fe_elem_off: 2b
363094 253: index:10262 entry: 200098d4 fa: 20007160 fe_elem_off: 2a
363096 254: index:8819 entry: 200098f0 fa: 20007154 fe_elem_off: 8
363097 255: index:10262 entry: 2000990c fa: 20007160 fe_elem_off: 2a  <------------ Duplicate
363098 256: index:0 entry: 20009928 fa: 0 fe_elem_off: 0
363099 257: index:0 entry: 20009944 fa: 0 fe_elem_off: 0
363100 258: index:0 entry: 20009960 fa: 0 fe_elem_off: 0
363102 259: index:0 entry: 2000997c fa: 0 fe_elem_off: 0
363103 260: index:0 entry: 20009998 fa: 0 fe_elem_off: 0
```
Now duplicate entry is present in absolute bookmark space (previously it was not the case)
This time measure walk time from index 50010 which also is present in sector bookmarks
```
compat> log -i 50010 -n 2 -t test_log
571519 Log test_log 2 entries walked in 32 ms
compat> log -b test_log
685500 0: index:0 entry: 20007d28 fa: 0 fe_elem_off: 0
685501 1: index:68837 entry: 20007d44 fa: 2000713c fe_elem_off: 2c
...
685611 88: index:50170 entry: 200086c8 fa: 2000791c fe_elem_off: 2b
685612 89: index:50010 entry: 200086e4 fa: 20007910 fe_elem_off: 2a
685614 90: index:49888 entry: 20008700 fa: 20007904 fe_elem_off: 2a
...
685821 253: index:10262 entry: 200098d4 fa: 20007160 fe_elem_off: 2a
685823 254: index:8819 entry: 200098f0 fa: 20007154 fe_elem_off: 8
685824 255: index:10262 entry: 2000990c fa: 20007160 fe_elem_off: 2a <--- Both were present in sector bookmarks
685825 256: index:50010 entry: 20009928 fa: 20007910 fe_elem_off: 2a <---
685827 257: index:0 entry: 20009944 fa: 0 fe_elem_off: 0
685828 258: index:0 entry: 20009960 fa: 0 fe_elem_off: 0
685829 259: index:0 entry: 2000997c fa: 0 fe_elem_off: 0
685830 260: index:0 entry: 20009998 fa: 0 fe_elem_off: 0

```
Walk is fast as expected
Next few walks without hitting sector bookmark
```
compat> log -i 30000 -n 2 -t test_log
926885 Log test_log 2 entries walked in 528 ms
compat> log -i 20000 -n 2 -t test_log
964547 Log test_log 2 entries walked in 252 ms
compat> log -i 10000 -n 2 -t test_log
969795 Log test_log 2 entries walked in 114 ms
compat> log -i 15000 -n 2 -t test_log
979476 Log test_log 2 entries walked in 117 ms
compat> log -i 11000 -n 2 -t test_log
986830 Log test_log 2 entries walked in 21 ms
compat> log -i 12000 -n 2 -t test_log
994202 Log test_log 2 entries walked in 26 ms
compat> log -i 13000 -n 2 -t test_log
997642 Log test_log 2 entries walked in 29 ms
compat> log -b test_log
1102179 0: index:0 entry: 20007d28 fa: 0 fe_elem_off: 0
1102180 1: index:68837 entry: 20007d44 fa: 2000713c fe_elem_off: 2c
...
1102506 254: index:8819 entry: 200098f0 fa: 20007154 fe_elem_off: 8
1102507 255: index:11000 entry: 2000990c fa: 20007160 fe_elem_off: 8cc
1102508 256: index:12000 entry: 20009928 fa: 2000716c fe_elem_off: 350
1102510 257: index:13000 entry: 20009944 fa: 2000716c fe_elem_off: ea9
1102511 258: index:20000 entry: 20009960 fa: 200071a8 fe_elem_off: 882
1102513 259: index:10000 entry: 2000997c fa: 20007154 fe_elem_off: cc8
1102514 260: index:15000 entry: 20009998 fa: 20007184 fe_elem_off: a5
```
Times seems to vary depending on index, grater index -> slower walk
Now all absolute bookmarks are there plus 255 sector ones
Measure time for walking 2 entries starting from index 50000
```
compat> stat spiflash
1228621 read_count: 19885
1228621 write_count: 4
1228622 erase_count: 0
1228622 error_count: 0
1228623 read_bytes: 215750
1228623 written_bytes: 157
compat> log -i 50000 -n 2 -t test_log
1256664 Log test_log 2 entries walked in 5869 ms
compat> stat spiflash
1259963 read_count: 96353
1259963 write_count: 4
1259964 erase_count: 0
1259964 error_count: 0
1259965 read_bytes: 1159963
1259965 written_bytes: 157
```
And here is measurements when sector bookmarks are off, this is single session
```
compat> log -i 50000 -n 2 -t test_log
Log test_log 2 entries walked in 67 ms
compat> log -i 10000 -n 2 -t test_log
030365 Log test_log 2 entries walked in 113 ms
compat> log -i 20000 -n 2 -t test_log
033948 Log test_log 2 entries walked in 101 ms
compat> log -i 30000 -n 2 -t test_log
037504 Log test_log 2 entries walked in 101 ms
compat> log -i 40000 -n 2 -t test_log
041646 Log test_log 2 entries walked in 80 ms
compat> log -i 10000 -n 2 -t test_log
046245 Log test_log 2 entries walked in 114 ms
compat> log -i 50000 -n 2 -t test_log
054318 Log test_log 2 entries walked in 66 ms
compat> log -i 10000 -n 2 -t test_log
061018 Log test_log 2 entries walked in 114 ms
compat> log -i 11000 -n 2 -t test_log
065112 Log test_log 2 entries walked in 104 ms
compat> log -i 13000 -n 2 -t test_log
067719 Log test_log 2 entries walked in 120 ms
compat> log -i 14000 -n 2 -t test_log
070770 Log test_log 2 entries walked in 110 ms
compat> log -i 15000 -n 2 -t test_log
073969 Log test_log 2 entries walked in 83 ms
compat> log -i 16000 -n 2 -t test_log
078615 Log test_log 2 entries walked in 107 ms
compat> log -i 50000 -n 2 -t test_log
083399 Log test_log 2 entries walked in 66 ms
```

Without sector bookmarks (absolute bookmarks are enabled) timing seems reasonable.

vrahane commented:
@kasjer I do not see the issues you mentioned above with my setup with default number of absolute bookmarks and using the sector count for sector bookmarks. Attaching the coolterm output. 
[CoolTerm Capture 2025-02-27 13-28-04.txt](https://github.com/user-attachments/files/19016511/CoolTerm.Capture.2025-02-27.13-28-04.txt)

kasjer commented:
@vrahane in your logs bookmarks look ok, in my case first entry looks strange:
```
1102179 0: index:0 entry: 20007d28 fa: 0 fe_elem_off: 0
```
This strange element at the beginning is always there. I have not checked why it's there but it probably could explain why sector bookmarks are not used

vrahane commented:
@kasjer it seems `fcbp->f_sector_cnt = sector_count - 1;` might be causing an issue. In my FCB init function, its set to the sector count.

vrahane commented:
I also tried with 6 absolute bookmarks. It works fine.

kasjer commented:
> @kasjer it seems `fcbp->f_sector_cnt = sector_count - 1;` might be causing an issue. In my FCB init function, its set to the sector count.

Yes, after correction to sector_count measured times are fast as expected

vrahane commented:
I am curious what timings do you see on your device ?

kasjer commented on sys/log/full/src/log_shell.c:
```diff
@@ -251,6 +276,29 @@ shell_log_dump_cmd(int argc, char **argv)
+#if MYNEWT_VAL(LOG_FCB)
+                console_printf("%u: index:%lu entry: %x fa: %x fe_elem_off: %lx\n", i,
+                               bmarks[i].lfb_index,
+                               (uintptr_t)&bmarks[i].lfb_entry,
```
printing addresses of each element of the table is not really useful.

kasjer commented on sys/log/full/src/log_shell.c:
```diff
@@ -251,6 +276,29 @@ shell_log_dump_cmd(int argc, char **argv)
+                console_printf("%u: index:%lu entry: %x fa: %x fe_elem_off: %lx\n", i,
+                               bmarks[i].lfb_index,
+                               (uintptr_t)&bmarks[i].lfb_entry,
+                               (uintptr_t)bmarks[i].lfb_entry.fe_area,
```
more useful would be to print offset in flash instead of address of array element

kasjer replied:
Updated version will not compile due `.` instead of `->`

vrahane commented on sys/log/full/src/log_fcb_bmark.c:
```diff
@@ -61,74 +195,256 @@ log_fcb_rotate_bmarks(struct fcb_log *fcb_log)
+
+        if (bset->lfs_non_sect_size < MYNEWT_VAL(LOG_FCB_NUM_ABS_BOOKMARKS)) {
+            bset->lfs_non_sect_size++;
+            bset->lfs_size++;
```
Hard to understand what you mean. Can you please recreate the above issue you are mentioning and paste the output.

kasjer replied:
Fresh output from log -b (only few last entries are shown here) 4 ABS bookmarks enabled
```
...
001165 186: index:252 fa: 103000 fe_elem_off: 4d
001166 187: index:155 fa: 102000 fe_elem_off: 6c
001166 188: index:108 fa: 101000 fe_elem_off: 76
001167 189: index:62 fa: 100000 fe_elem_off: 8
001167 190: unused
001167 191: unused
001168 192: unused
001168 193: unused
```
now few walks, 3 of them for same index, 4th for different index
```
compat> log -i 62 -n 2 -t test_log
005561 Log test_log 2 entries walked in 0 ms
compat> log -i 62 -n 2 -t test_log
005774 Log test_log 2 entries walked in 7 ms
compat> log -i 62 -n 2 -t test_log
006266 Log test_log 2 entries walked in 0 ms
compat> log -i 108 -n 2 -t test_log
007396 Log test_log 2 entries walked in 46 ms
```
now `log -b test_log` again
```
007771 187: index:155 fa: 102000 fe_elem_off: 6c
007772 188: index:108 fa: 101000 fe_elem_off: 76
007772 189: index:62 fa: 100000 fe_elem_off: 8
007773 190: index:62 fa: 100000 fe_elem_off: 8
007773 191: unused
007773 192: unused
007774 193: index:108 fa: 101000 fe_elem_off: 76
```
At this point `lfs_size` and `lfs_non_sect_size` were incremented even though bookmark was not added.
`lfs_next_non_sect` was also modified leaving spaces in bookmark array

vrahane replied:
Thanks, that helps, fixing it.

vrahane commented on sys/log/full/src/log_fcb_bmark.c:
```diff
@@ -61,74 +195,256 @@ log_fcb_rotate_bmarks(struct fcb_log *fcb_log)
+
+    if (bset->lfs_size < fcb_log->fl_bset.lfs_sect_cap) {
+        bset->lfs_size++;
+        bset->lfs_next_sect = (bset->lfs_next_sect - 1) %
+                              fcb_log->fl_bset.lfs_sect_cap;
```
So, are you suggesting I remove this ?

kasjer replied:
In code you implemented, correct way to deal with such case would be to have:
```c
bset->lfs_next_sect = fcb_log->fl_bset.lfs_sect_cap - 1;
```
when `lfs_next_sect` is already 0 and decrement it when it's not

vrahane replied:
`log_fcb_init_sector_bmarks()` adds the bookmarks. It initializes `bset->lfs_next_sect = bset->lfs_sect_cap - 1;` Check line `log_fcb_bmark.c: 58`. `mod` helps `bset->lfs_next_sect` decrement correctly. If/Else is not required. Edge cases explained below:

### Edge Case 1: Initial Value at Maximum
If `bset->lfs_next_sect` starts at the maximum value of `fcb_log->fl_bset.lfs_sect_cap - 1` (which is 4), the sequence of values will be:

- Initial value: 4
- After first iteration: (4 - 1) % 5 = 3
- After second iteration: (3 - 1) % 5 = 2
- After third iteration: (2 - 1) % 5 = 1
- After fourth iteration: (1 - 1) % 5 = 0
- After fifth iteration: (0 - 1) % 5 = 4 (since -1 % 5 = 4 in unsigned arithmetic)

### Edge Case 2: Initial Value at Zero
If `bset->lfs_next_sect` starts at 0, the sequence of values will be:

- Initial value: 0
- After first iteration: (0 - 1) % 5 = 4 (since -1 % 5 = 4 in unsigned arithmetic)
- After second iteration: (4 - 1) % 5 = 3
- After third iteration: (3 - 1) % 5 = 2
- After fourth iteration: (2 - 1) % 5 = 1
- After fifth iteration: (1 - 1) % 5 = 0

### Edge Case 3: Large Initial Value
If `bset->lfs_next_sect` starts at a large value, say 10, the sequence of values will be:

- Initial value: 10
- After first iteration: (10 - 1) % 5 = 9 % 5 = 4
- After second iteration: (4 - 1) % 5 = 3
- After third iteration: (3 - 1) % 5 = 2
- After fourth iteration: (2 - 1) % 5 = 1
- After fifth iteration: (1 - 1) % 5 = 0

### Edge Case 4: Negative Initial Value
Since `bset->lfs_next_sect` is a `uint32_t`, it cannot be negative. However, if we consider the behavior of the modulo operation with negative numbers in general, it would wrap around to a positive value within the range of `fcb_log->fl_bset.lfs_sect_cap`.

kasjer replied:
## Nice explanation, except it's not correct.

I guess you never actually tried how it would work
Here is test code that shows how this modulo logic behaves.
Since error in code is when `lfs_next_sect` is decremented from 0, following code prints 10 values of lfs_next_sect starting from 4 to see what sequence of lfs_next_sect would look like.
Starting from capacity 5 as in your example
```c
    uint32_t lfs_next_sect;
    int lfs_sect_cap[] = {5, 7, 8, 192, 256};

    for (int i = 0; i < 5; ++i) {
        /* Let start from 4 and continue towards expected rollover */
        lfs_next_sect = 4;
        printf("Test when capacity is %d\n", lfs_sect_cap[i]);
        for (int j = 0; j < 10; ++j) {
            lfs_next_sect = (lfs_next_sect - 1) % lfs_sect_cap[i];
            printf("lfs_next_sec = %lu\n", lfs_next_sect);
        }
    }
```
and here is output from this code
## Most interesting result is from capacity 5 as per your example

### Test when capacity is 5
```
lfs_next_sec = 3
lfs_next_sec = 2
lfs_next_sec = 1
lfs_next_sec = 0
lfs_next_sec = 0
lfs_next_sec = 0
lfs_next_sec = 0
lfs_next_sec = 0
lfs_next_sec = 0
lfs_next_sec = 0
```
### Test when capacity is 7
```
lfs_next_sec = 3
lfs_next_sec = 2
lfs_next_sec = 1
lfs_next_sec = 0
lfs_next_sec = 3
lfs_next_sec = 2
lfs_next_sec = 1
lfs_next_sec = 0
lfs_next_sec = 3
lfs_next_sec = 2
```
### Test when capacity is 8
```
lfs_next_sec = 3
lfs_next_sec = 2
lfs_next_sec = 1
lfs_next_sec = 0
lfs_next_sec = 7
lfs_next_sec = 6
lfs_next_sec = 5
lfs_next_sec = 4
lfs_next_sec = 3
lfs_next_sec = 2
```
### Test when capacity is 192
```
lfs_next_sec = 3
lfs_next_sec = 2
lfs_next_sec = 1
lfs_next_sec = 0
lfs_next_sec = 63
lfs_next_sec = 62
lfs_next_sec = 61
lfs_next_sec = 60
lfs_next_sec = 59
lfs_next_sec = 58
```
### Test when capacity is 256
```
lfs_next_sec = 3
lfs_next_sec = 2
lfs_next_sec = 1
lfs_next_sec = 0
lfs_next_sec = 255
lfs_next_sec = 254
lfs_next_sec = 253
lfs_next_sec = 252
lfs_next_sec = 251
lfs_next_sec = 250
```
computation is only valid when capacity is power of 2

vrahane replied:
Thanks for the run down. Fixed it.

kasjer replied:
The point of using modulo operator for index rollover is that there is no condition check involved. So one liner can do the job.
In current version there is condition check and modulo instead of
```c
bset->lfs_next_sect--;
if (bset->lfs_next_sect >= fcb_log->fl_bset.lfs_sect_cap) {
    bset->lfs_next_sect = fcb_log->fl_bset.lfs_sect_cap - 1;
}
```

vrahane commented on sys/log/full/src/log_fcb_bmark.c:
```diff
@@ -22,17 +22,151 @@
+#if MYNEWT_VAL(LOG_FCB)
+        j = 0;
+
+        /* Keep skipping sectors until lfs_sect_bmark_itvl is reached */
+        do {
+            fa = fcb_getnext_area(&fcb_log->fl_fcb, loc.fe_area);
+            if (!fa) {
+                break;
+            }
+            j++;
+        } while (j < bset->lfs_sect_bmark_itvl);
```
Thanks, good catch.

vrahane commented on sys/log/full/src/log_fcb_bmark.c:
```diff
@@ -22,17 +22,151 @@
+int
 log_fcb_init_bmarks(struct fcb_log *fcb_log,
-                    struct log_fcb_bmark *buf, int bmark_count)
+                    struct log_fcb_bmark *buf, int avl_bmark_cnt,
```
It is for understanding the code better, the comparison between required bookmark count and available bookmark count is performed below. Hence, the name change. I have updated the description to reflect the same and added a comment as well.

vrahane commented on sys/log/full/src/log_fcb_bmark.c:
```diff
@@ -22,17 +22,151 @@
 {
-    fcb_log->fl_bset = (struct log_fcb_bset) {
+    struct log_fcb_bset *bset = &fcb_log->fl_bset;
+    int reqd_bmark_cnt = MYNEWT_VAL(LOG_FCB_NUM_ABS_BOOKMARKS);
```
It is for understanding the code better, the comparison between required bookmark count and available bookmark count is performed below. Hence, the name change. I will update the description to reflect the same and added a comment as well.

vrahane commented on sys/log/full/src/log_fcb_bmark.c:
```diff
@@ -22,17 +22,151 @@
+        return SYS_EINVAL;
+    }
+
+    memset(bset, 0, sizeof(*bset));
```
Thanks, I moved it around a bit, so this became useless.

vrahane commented on sys/log/full/src/log_fcb_bmark.c:
```diff
@@ -22,17 +22,151 @@
+    }
+#endif
+
+#if MYNEWT_VAL(LOG_FCB_SECTOR_BOOKMARKS)
+    if (en_sect_bmarks) {
```
Good catch, thanks.

kasjer replied:
Updated version removed double preprocessor section but unnecessary second `if` statement is still there. `en_sect_bmarks` is not change in the function there is not need to check it twice.

vrahane commented on sys/log/full/include/log/log_fcb.h:
```diff
@@ -125,39 +157,46 @@ void log_fcb_init_bmarks(struct fcb_log *fcb_log,
  */
-const struct log_fcb_bmark *
-log_fcb_closest_bmark(const struct fcb_log *fcb_log, uint32_t index);
+struct log_fcb_bmark *
```
As mentioned earlier in the PR. Trying to reduce usage of const because we are seeing a performance hit. I feel like other places can have consts, for example log_headers or FCB entry, etc since thats part of the logging code and APIs already but bookmarks on their own don't need to be consts.

kasjer replied:
I think that you can create separate PR that drops const in places that you can show change in performance due to const removal.
Removing const from random places in already complicated code make things harder to review.
Right now I'm not convinced that this change is needed.

vrahane replied:
I am already changing this code so much, I strongly feel this change belongs in this PR. If it was an unrelated PR I would open one separately.

kasjer replied:
Very well, you may just start gathering evidence that this change improves performance

vrahane replied:
And who are you to ask me for evidence ? Its not like you give evidence for every line of code you write. The most recent one for example was changed without any supporting evidence: https://github.com/apache/mynewt-core/pull/3381  So, please stop being so dismissive. If you continue being a pain, I will escalate this. @sjanc

vrahane commented on sys/log/full/src/log_fcb_bmark.c:
```diff
@@ -61,74 +195,256 @@ log_fcb_rotate_bmarks(struct fcb_log *fcb_log)
+        return SYS_ENOMEM;
+    }
+
+    en_sect_bmark = sect_bmark & bset->lfs_en_sect_bmarks;
+    (void)en_sect_bmark;
```
`en_sect_bmark` was earlier used in both cases, so It was left that way, removing it. Makes sense.

kasjer commented on sys/log/full/src/log_fcb_bmark.c:
```diff
@@ -61,74 +196,245 @@ log_fcb_rotate_bmarks(struct fcb_log *fcb_log)
+     */
     for (i = 0; i < fcb_log->fl_bset.lfs_size; i++) {
         bmark = &fcb_log->fl_bset.lfs_bmarks[i];
+#if MYNEWT_VAL(LOG_FCB)
```
This loop used to work before sector bookmarks were added.
Now if space for sector bookmarks is not fully filled no sector bookmark is found.
Depending on how many sector bookmarks are actually filled non-sector bookmark are most likely not found either.

All this is due to reverse order of insertion of sector bookmarks and forward search of the array.
Use lfs_size to limit loop results in ignoring non sector bookmarks even it they were already used.

vrahane replied:
So, the way it works is `lfs_sect_cap` gets set to the appropriate number of sector bookmarks that can fit in the available memory thats provided. If there is more memory allocated, bookmarks for the number of sectors will get allocated at the beginning. After the sector bookmarks without any empty space, non-sector bookmarks will get allocated. So, there is no gap. Do you think this is reproducible ? If so, can you please provide output of the same with an example.

kasjer replied:
To see that this does not work simply clear half of you logs then reboot. Or clear all logs and generate only some entries so log is not full.
Then measure log walk starting from any index that is in sector bookmarks.
If everything works fine, after displaying log starting from index that is present in bookmarks there should not be any additional entry in non-sector bookmarks.
Here is output from log that has 333 sectors and 6 absolute bookmark space
```
...
002870 327: index:493 fa_off: 26e000 fe_elem_off: 32
002870 328: index:396 fa_off: 26d000 fe_elem_off: 32
002870 329: index:299 fa_off: 26c000 fe_elem_off: 32
002871 330: index:202 fa_off: 26b000 fe_elem_off: 32
002871 331: index:105 fa_off: 26a000 fe_elem_off: 32
002871 332: index:8 fa_off: 269000 fe_elem_off: 8
002871 333: index:0 fa_off: 20000311 fe_elem_off: 0
002872 334: index:0 fa_off: 20000311 fe_elem_off: 0
002872 335: index:0 fa_off: 20000311 fe_elem_off: 0
002873 336: index:0 fa_off: 20000311 fe_elem_off: 0
002873 337: index:0 fa_off: 20000311 fe_elem_off: 0
002873 338: index:0 fa_off: 20000311 fe_elem_off: 0
compat> log -i 105 -n 2 -t datarec_dbg
009706 Log datarec_dbg 2 entries walked in 46 ms
compat> log -b datarec_dbg
...
010990 327: index:493 fa_off: 26e000 fe_elem_off: 32
010990 328: index:396 fa_off: 26d000 fe_elem_off: 32
010991 329: index:299 fa_off: 26c000 fe_elem_off: 32
010991 330: index:202 fa_off: 26b000 fe_elem_off: 32
010991 331: index:105 fa_off: 26a000 fe_elem_off: 32
010992 332: index:8 fa_off: 269000 fe_elem_off: 8
010992 333: index:105 fa_off: 26a000 fe_elem_off: 32
010992 334: index:0 fa_off: 20000311 fe_elem_off: 0
010993 335: index:0 fa_off: 20000311 fe_elem_off: 0
010993 336: index:0 fa_off: 20000311 fe_elem_off: 0
010994 337: index:0 fa_off: 20000311 fe_elem_off: 0
010994 338: index:0 fa_off: 20000311 fe_elem_off: 0
compat> log -i 105 -n 2 -t datarec_dbg
012411 Log datarec_dbg 2 entries walked in 46 ms
```
Absolute bookmark is added here even though it's duplicate of existing sector bookmark.
When second walk starts, no bookmark is found due to how lfs_size is used to limit number of iterations.

vrahane replied:
Fixed.

vrahane commented on sys/log/full/src/log_fcb.c:
```diff
@@ -209,13 +215,21 @@ log_fcb_start_append(struct log *log, int len, struct fcb_entry *loc)
     struct flash_area *old_fa;
     int rc = 0;
+#if MYNEWT_VAL(LOG_FCB_SECTOR_BOOKMARKS)
+    int active_sector_cnt = 0;
```
I was using active_id earlier but the name confused me into thinking I did something wrong. So, I changed it.

vrahane replied:
Ok, using sector ID now, I tried it out and it seems to work fine. Let me know if you see any issues with it.

kasjer replied:
Comment `/* Cache sector count before appending */` is still misleading.

There is no point in initializing `active_id` twice. (same for `idx`)

Considering that `f_active_id` is uint16_t, condition
```c
if (fcb->f_active_id > active_id) {
```
probably would be more future proof this way:
```c
if (fcb->f_active_id != active_id) {
```

vrahane replied:
I don't how this change will make it future proof, but I am going to change it.

vrahane commented:
Note: The style check incorrectly flags the unit test functions. It does not understand that its a function, it thinks its a code block.

sjanc commented:
> Note: The style check incorrectly flags the unit test functions. It does not understand that its a function, it thinks its a code block.

yes, there are couple of uncrustify config tuning needed that I haven't figure out yet (especially those related to macros as in here).   So we hit false-positive with this from time to time...
